openc3/lib/openc3/utilities/local_mode.rb

```diff
@@ -0,0 +1,507 @@
+# encoding: ascii-8bit
+
+# Copyright 2022 OpenC3, Inc.
+# All Rights Reserved.
+#
+# This program is free software; you can modify and/or redistribute it
+# under the terms of the GNU Affero General Public License
+# as published by the Free Software Foundation; version 3 with
+# attribution addendums as found in the LICENSE.txt
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+
+require 'fileutils'
+require 'json'
+# require 'openc3/models/gem_model' # These are used by also create circular dependency
+# require 'openc3/models/plugin_model' # These are used by also create circular dependency
+require 'openc3/utilities/s3'
+
+module OpenC3
+  module LocalMode
+    OPENC3_LOCAL_MODE_PATH = ENV['OPENC3_LOCAL_MODE_PATH'] || "/plugins"
+
+    DEFAULT_PLUGINS = [
+      'openc3-tool-admin',
+      'openc3-tool-autonomic',
+      'openc3-tool-base',
+      'openc3-tool-calendar',
+      'openc3-tool-cmdsender',
+      'openc3-tool-cmdtlmserver',
+      'openc3-tool-dataextractor',
+      'openc3-tool-dataviewer',
+      'openc3-tool-handbooks',
+      'openc3-tool-limitsmonitor',
+      'openc3-tool-packetviewer',
+      'openc3-tool-scriptrunner',
+      'openc3-tool-tablemanager',
+      'openc3-tool-tlmgrapher',
+      'openc3-tool-tlmviewer',
+      'openc3-enterprise-tool-base',
+      'openc3-enterprise-tool-admin',
+    ]
+
+    # Install plugins from local plugins folder
+    def self.local_init
+      if ENV['OPENC3_LOCAL_MODE'] and Dir.exist?(OPENC3_LOCAL_MODE_PATH)
+        puts "Local init running: #{OPENC3_LOCAL_MODE_PATH} exists"
+        Dir.each_child(OPENC3_LOCAL_MODE_PATH).each do |scope_dir|
+          next unless File.directory?("#{OPENC3_LOCAL_MODE_PATH}/#{scope_dir}")
+          puts "Local init found scope: #{scope_dir}"
+          Dir.each_child("#{OPENC3_LOCAL_MODE_PATH}/#{scope_dir}") do |plugin_dir|
+            full_folder_path = "#{OPENC3_LOCAL_MODE_PATH}/#{scope_dir}/#{plugin_dir}"
+            next if plugin_dir == "targets_modified" or not File.directory?(full_folder_path)
+            puts "Local init found plugin_dir: #{full_folder_path}"
+            gems, plugin_instance = scan_plugin_dir(full_folder_path)
+
+            if gems.length > 1
+              puts "Local plugin folder contains more than one gem - skipping: #{full_folder_path}"
```

Comment:
When is this possible? If you require external gems in your plugin?

Reply:
Its a user controlled folder - anything is possible.